### PR TITLE
Prevent DataPipe attributes that use existing functional datapipe names

### LINKED ISSUE #101: weight initializations for conv2d and linear
Allow different weight initialization schemes for conv2d and linear.
See: https://github.com/Kaixhin/nninit for some schemes.

### LINKED ISSUE #101: Prevent (or at least flag) DataPipe attributes and methods that use existing functional datapipe names
## 🚀 Feature
We should have some checks/tests that flag when a DataPipe has an attribute/method that shares 

For example, `Batcher` (or `BatcherIterDataPipe`) has the functional datapipe name `batch`. However, currently there is nothing to prevent other IterDataPipes to use `batch` as the name of an attribute or method.

The change in the following PR is a [good example](https://github.com/pytorch/pytorch/pull/67374#discussion_r738820749).

## Motivation

If this feature is not implemented, then a DataPipe can have multiple attributes/methods with the same name, potentially causing confusion and bugs.

## Alternatives

Ideally, we should be able to flag this issue during development (within IDEs).

If we cannot automatically prevent this during development, we can have a check in `register_datapipe_as_function` or a CI check that ensures all attributes and methods are compliant.

`mypy` also should be able to flag this issue if the `.pyi` file has a complete set of method interfaces for all built-in DataPipes (including those in TorchData). This becomes trickier for user-defined DataPipes.

### PR DESCRIPTION
**Pull request by ejguan**
Partially fixes [#101](https://github.com/pytorch/data/issues/101)

If users want to use a method to override functional API, I don't know we have a simple way to prevent such behavior unless we register functional API into metaclass.

This may need to be switched to warning since this is a BC breaking change

### CODE REVIEW

File: test/test_datapipe.py [hunk 1]
```diff
@@ -274,6 +274,18 @@ def tearDown(self):
         except Exception as e:
             warnings.warn("TestIterableDatasetBasic was not able to cleanup temp dir due to {}".format(str(e)))
 
+    def test_datapipe_setattr(self):
```
Comment (NivekT) [inline test/test_datapipe.py:277]: nit: Maybe duplicate this for `MapDataPipe` too just for completeenss?

**Review by NivekT — APPROVED** [review]
Thanks for coming up with this solution!

This is definitely BC-breaking but I feel like throwing an error is better than silent failure.

cc: @VitalyFedyunin

**Comment by VitalyFedyunin** [comment]
It is not only BC breaking, but potentially more confusing. Imagine someone imported let's say pandas-datapipes, and immediately all current DataPipes which had pandas as attributes will stop working.  In general, it is much easier to check if my current DP has some attributes that prevent `.foo` from working in comparison to searching all dependencies for a weird override.

**Comment by VitalyFedyunin** [comment]
However, this problem remains valid. How about we add a test that will verify all data pipes attributes for a list of known functional data pipes. This test should be recommended for all DP contributions similar to the Serializability test.

**Comment by ejguan** [comment]
> Imagine someone imported let's say pandas-datapipes, and immediately all current DataPipes which had pandas as attributes will stop working. In general, it is much easier to check if my current DP has some attributes that prevent `.foo` from working in comparison to searching all dependencies for a weird override.

I am not sure what do you mean. For pandas-datapipes, if users assign a few methods to this datapipe, the current DataPipes won't break because these methods are attached to `pandas-datapipes` rather than stored into `IterDataPipe.functions`.

For the case that users explicitly want to override a few attributes, we should provide such option. If we want to do so, I may need to add an argument to the metaclass to let users determine if they want to override.

**Pull request closed without merging**